### ISSUE TITLE
\n randomly added in tree mode

### ISSUE THREAD

**Issue by Didjynn** [josdejong/jsoneditor#936]
\n are randomly added when I try to paste text when I am in tree mode.
It happens in the left and right box but only in tree mode

**Comment by josdejong** [comment]
A return character is no valid character in JSON, therefore JSONEditor escapes returns into `\n`. Does that explain it?

See also #141, #137.

**Comment by Didjynn** [comment]
nope, it replaces spaces and sometimes  ' .

I can get two at one place. I have no idea what is causing this but it doesn not do that if I paste the text in code mode then I go to tree mode.

**Comment by josdejong** [comment]
Hm, I'm not sure what's going on. Can it be that you're pasting formatted text from Word or something like that? (If so, it should not happen when copying something from a plain text editor).

Is there a way I can reproduce the issue you are encountering?

**Comment by Didjynn** [comment]
I don't know it's text from open office writer.

I just tried to transform it into a pdf and I tried with the same text but copied from the pdf this time. Each \n is a new line but not manually made, it's just the pdf going to a new line. That's quite annoying, is there a way to get rid of that ?

**Comment by Didjynn** [comment]
I don't know it's text from open office writer.

I just tried to transform it into a pdf and I tried with the same text but copied from the pdf this time. Each \n is a new line but not manually made, it's just the pdf going to a new line. That's quite annoying, is there a way to get rid of that ?

**Comment by Didjynn** [comment]
so, does it mean i can't copy text from microsoft word, pdf or open office writer ? Are they "plain text editors" ?

**Comment by josdejong** [comment]
Copy/paste text from any application should just work. It is normal though that if you paste text with multiple lines, that the editor replaces this with escaped `\n` characters - else it is no valid JSON. Other special characters that are no valid JSON will also be replaced.

Is there a way I can reproduce your issue? Without that I can't do more than guessing what's going on on your screen.

**Comment by Didjynn** [comment]
As I said, it works if you copy / paste text from open office writer or from a pdf. I wonder if there isn't other symbols transformed in \n.

for example this text :
Les jeux de gestion, de stratégie et les RPG que l'on trouve dans votre ligne éditoriale à Goblinz tudio sont des productions dans lesquelles je me retrouve. Ils ont en commun avec l'entreprise et plus particulièrement les métiers de Producer l'organisation, la planification et l'optimisation.

Becomes :

\n\nLes\njeux de gestion, de stratégie et les RPG que l'on trouve dans votre\nligne éditorialeà Goblinz Studio sont des\nproductions dans lesquelles je me retrouve. Ils ont en commun avec\nl'entreprise et plus particulièrement les métiers de Producer\nl'organisation, la planification et\nl'optimisation.\n\n

So there is a double \n\n at the beginning and ending of every text copied. Also, at the beginnig, there is a \n between "Les" and "jeux". There is no new line to the \n is replacing a blank space. The same happen here : votre\nligne
And I have no idea why.

**Comment by josdejong** [comment]
Thanks for the example. Can you share a (minimal) Word document with this text? When I copy it from this github issue, I cannot reproduce getting a `\n\n` at the start or having a `\n` placed in `Les Jeux`.

**Comment by Didjynn** [comment]
I hope you'll get the file if I answer by mail because I can't send it
directly on github, the file's format isn't supported. I send 2 files: one
is open office writer, the other was saved as a word document. Both make
the \n happen the same way.

Le dim. 29 mars 2020 à 17:15, Jos de Jong <notifications@github.com> a
écrit :

**Comment by Didjynn** [comment]
did you receive the mail response ?

**Comment by josdejong** [comment]
No :(

**Comment by Didjynn** [comment]
and this time ?

**Comment by josdejong** [comment]
Thanks I received the files on my email address. Will look into it when I have some time.

**Comment by josdejong** [comment]
I tried out copying the text from your test documents into a field in JSONEditor. In my case the editor shows an error "Cannot parse value into JSON" and simply refuses to put the text in the field. 

When I paste the text in a plain text editor (like Notepad), and then I copy and paste the plain text into a field in JSONEditor, it all works. I'm not sure yet what's going on here, maybe some character encoding thing or so. To be continued... :male_detective: 

I've attached a zip file with your test documents (zip is supported): [test.zip](https://github.com/josdejong/jsoneditor/files/4432196/test.zip)

**Comment by Didjynn** [comment]
o yes, that's weird. Both documents are working on my side but maybe it comes from the fact I open them and copy them from open office writer.

Did you send these files for me to do something with them ?

**Comment by josdejong** [comment]
> Did you send these files for me to do something with them ?

No just so my future self or anyone interested in debugging this can find them.

**Comment by Didjynn** [comment]
Hello,
I was thinking about it and how you could not manage to reproduce the bug the same way I have it.

Maybe if you have time the best way to do so would be to install open office writer on your computer, write anything in it and copy paste it in tree mode ? (on the website https://jsoneditoronline.org/)

Because I have this really 100% of the time contrary to the "reverse writing" bug. That does not happen if I copy paste from firefox for example. In the worst case, an option  to automatically clean the \n or to prevent them from being created after pasting would be welcome to me

Cheers

**Comment by josdejong** [comment]
I opened your documents in LibreOffice. Are you using LibreOffice or OpenOffice?

**Comment by Didjynn** [comment]
OpenOffice Writer

**Comment by josdejong** [comment]
I managed to reproduce the issue 🎉 

The text copied from OpenOffice contains paragraphs, like:

```html
<p>Hello world</p>
<p>test paste from OpenOffice</p>
<p><br></p>
<p>test</p>
```

When pasting text with HTML formatting, JSONEditor uses HTML `div.innerText` to get the plain text equivalent. However, the `innerText` of this contains too many returns: 

``` 
"Hello world\n\ntest paste from OpenOffice\n\n\n\n\ntest"
```

Here is a demo: https://jsbin.com/begicis/edit?html,output

Still have to figure out why this is.

**Comment by josdejong** [comment]
Solution is probably to just use `.textContent` instead of `.innerText`, see:

https://stackoverflow.com/questions/35238840/inner-text-not-working-for-paragraph-and-heading-html-elements

**Comment by josdejong** [comment]
This is fixed now in `jsoneditor@8.6.8` and web application `v5.5.4`. Can you give it a try @Didjynn ?

**Comment by Didjynn** [comment]
That would be great since I didn't find a way to change open office's behavior to avoid this.

I made some tries and I seem to find 2 bugs:
- First time I pasted the text I got the same \n thing happening. I reloaded the page just in case and this time I did not get the \n. However on some tries I managed to make this \n happen again. But I am not able to reproduce it so I'll keep you in touch about that.
- It seems I can't click (in tree mode) in a text area I just changed if I click out of it or press the tab key to "validate" it. It even seem to be more than only the last one and keep a kind of memory of each ones I changed. This one seem to happen 100% of the time. I think it is linked to only text pasted from open office since I tried with some text copied from firefox and it doesn't do this.

I appreciate to see this json editor becoming better, thanks for your time

**Comment by Didjynn** [comment]
I also noted that if you clic on the right side of a "blocked" text area the writing cursor does this :

![image](https://user-images.githubusercontent.com/62567023/81825784-56c96300-9537-11ea-972e-ea7aa254ddab.png)

But it gives you back access to the blocked text area, as long as you act in it with the keyboard, not the mouse. The mouse seem blocked, it doesn't react like it should in this text area.

**Comment by josdejong** [comment]
Thanks for your feedback @Didjynn . Your first issue with the new behavior not yet working could be that your browser still had the old version of the application in its cache.

I did my best to reproduce the issue you describe but I did not manage to get the input field non-editable or with the cursor so strangely above the text. How can I reproduce that issue?

**Comment by Didjynn** [comment]
You need to cpy some text from open office writer, paste it in the test box. Then left click anywhere to validate (tab key works too).
You can try clicking on a new text field or a new key, it should work too.
This is it to block it with the mouse.

Then left click on the right side of the previously filled text box for the weird positionning.

I hope it helps

**Comment by josdejong** [comment]
hm sorry I can't get it reproduced here.

**Comment by Didjynn** [comment]
The weird cursor positionning doesn't seem to happen all the time but the blocked text area does. Maybe should I clean the firefox's cache ?

**Comment by josdejong** [comment]
I do not think that that kind of behavior can be caching issue. You could try if it is also an issue in other browser on your laptop, or try out if it can be related to certain browser extensions by opening the web application in incognito mode. What operating system and language are you using?
